packages/react-core/src/components/TreeView/TreeViewListItem.tsx

```diff
@@ -186,20 +198,21 @@ const TreeViewListItemBase: React.FunctionComponent<TreeViewListItemProps> = ({
               onClick={(evt: React.MouseEvent) => {
                 if (!hasCheck) {
                   onSelect && onSelect(evt, itemData, parentItem);
-                  if (children && evt.isDefaultPrevented() !== true) {
+                  if (!hasSelectableNodes && children && evt.isDefaultPrevented() !== true) {
                     setIsExpanded(!internalIsExpanded);
                   }
                 }
               }}
-              tabIndex={-1}
+              tabIndex={hasSelectableNodes ? 1 : -1}
+              {...((hasCheck || hasSelectableNodes) && { htmlFor: randomId, id: `label-${randomId}` })}
```

Comment:
`for` should only apply to `hasCheck` - ie, when the node is a `<label>`.

The `id` is good here, but technically doesn't need to apply to non-expandable selectable nodes. I think the case would be `hasCheck || (hasSelectableNodes && children)` to apply an `id`

Reply:
Updated. There will be a follow up to iron out the remaining keyboard issues.